File: app/src/main/java/com/example/autoflow/data/WorkflowEntity.java
package com.example.autoflow.data;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.Ignore;
import androidx.room.PrimaryKey;
import org.json.JSONException;
import org.json.JSONObject;

import com.example.autoflow.model.Action;
import com.example.autoflow.model.Trigger;
import com.example.autoflow.util.Constants;
import java.util.Objects;

@Entity(tableName = "workflows")
public class WorkflowEntity {
    @PrimaryKey(autoGenerate = true)
    public long id;

    @ColumnInfo(name = "workflow_name")
    @NonNull
    private String workflowName = "";

    @ColumnInfo(name = "is_enabled")
    private boolean isEnabled = false;

    @ColumnInfo(name = "trigger_details")
    @NonNull
    private String triggerDetails = "";

    @ColumnInfo(name = "action_details")
    @NonNull
    private String actionDetails = "";

    // Default constructor required by Room
    public WorkflowEntity() {
        this.workflowName = "";
        this.triggerDetails = "";
        this.actionDetails = "";
    }

    @Ignore
    public WorkflowEntity(@NonNull String workflowName, boolean isEnabled,
                          @NonNull String triggerDetails, @NonNull String actionDetails) {
        this.workflowName = Objects.requireNonNull(workflowName, "Workflow name cannot be null");
        this.isEnabled = isEnabled;
        this.triggerDetails = Objects.requireNonNull(triggerDetails, "Trigger details cannot be null");
        this.actionDetails = Objects.requireNonNull(actionDetails, "Action details cannot be null");
    }

    // Getters and Setters with null safety
    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    @NonNull
    public String getWorkflowName() {
        return workflowName != null ? workflowName : "";
    }

    public void setWorkflowName(@Nullable String workflowName) {
        this.workflowName = workflowName != null ? workflowName : "";
    }

    public boolean isEnabled() {
        return isEnabled;
    }

    public void setEnabled(boolean enabled) {
        this.isEnabled = enabled;
    }

    @NonNull
    public String getTriggerDetails() {
        return triggerDetails != null ? triggerDetails : "";
    }

    public void setTriggerDetails(@Nullable String triggerDetails) {
        this.triggerDetails = triggerDetails != null ? triggerDetails : "";
    }

    @NonNull
    public String getActionDetails() {
        return actionDetails != null ? actionDetails : "";
    }

    public void setActionDetails(@Nullable String actionDetails) {
        this.actionDetails = actionDetails != null ? actionDetails : "";
    }

    // In WorkflowEntity.java
    public static WorkflowEntity fromTriggerAndAction(
            @NonNull String workflowName,
            boolean isEnabled,
            @NonNull com.example.autoflow.model.Trigger trigger,  // ← Full package name
            @NonNull com.example.autoflow.model.Action action) {   // ← Full package name

        try {
            Log.d("WorkflowEntity", "🔵 fromTriggerAndAction called");

            // Convert Trigger to JSON
            JSONObject triggerJson = new JSONObject();
            triggerJson.put("type", trigger.type);
            triggerJson.put("value", trigger.value);
            String triggerDetails = triggerJson.toString();

            // Convert Action to JSON
            JSONObject actionJson = new JSONObject();
            actionJson.put("type", action.type);
            if (action.title != null) actionJson.put("title", action.title);
            if (action.message != null) actionJson.put("message", action.message);
            if (action.priority != null) actionJson.put("priority", action.priority);
            if (action.getValue() != null) actionJson.put("value", action.getValue());
            String actionDetails = actionJson.toString();

            // Create entity
            WorkflowEntity entity = new WorkflowEntity();
            entity.setWorkflowName(workflowName);
            entity.setTriggerDetails(triggerDetails);
            entity.setActionDetails(actionDetails);
            entity.setEnabled(isEnabled);

            Log.d("WorkflowEntity", "✅ WorkflowEntity created");
            return entity;

        } catch (Exception e) {
            Log.e("WorkflowEntity", "❌ Error", e);
            return null;
        }
    }


    @Nullable
    public com.example.autoflow.model.Trigger toTrigger() {
        // Null and empty string safety check
        if (triggerDetails == null || triggerDetails.trim().isEmpty()) {
            return null;
        }

        try {
            JSONObject json = new JSONObject(triggerDetails);

            // Extract values with null safety
            long id = json.optLong("id", 0L);
            long workflowId = json.optLong("workflowId", 0L);
            String type = json.optString("type");
            String value = json.optString("value", "");

            // Validate essential fields
            if (type == null || type.trim().isEmpty()) {
                System.err.println("Trigger type is null or empty, cannot create Trigger object");
                return null;
            }

            // Create trigger with null-safe values
            return new com.example.autoflow.model.Trigger(id, workflowId, type.trim(), value);

        } catch (JSONException e) {
            System.err.println("Error parsing trigger JSON: " + e.getMessage());
            return null;
        } catch (IllegalArgumentException e) {
            System.err.println("Invalid trigger data: " + e.getMessage());
            return null;
        } catch (Exception e) {
            System.err.println("Unexpected error creating trigger: " + e.getMessage());
            return null;
        }
    }

    @Nullable
    public com.example.autoflow.model.Action toAction() {
        // Null and empty string safety check
        if (actionDetails == null || actionDetails.trim().isEmpty()) {
            return null;
        }

        try {
            JSONObject json = new JSONObject(actionDetails);
            String type = json.optString("type");

            // Validate essential type field
            if (type == null || type.trim().isEmpty()) {
                System.err.println("Action type is null or empty, cannot create Action object");
                return null;
            }

            com.example.autoflow.model.Action action;
            type = type.trim();

            // Handle notification action with null safety
            if (Constants.ACTION_SEND_NOTIFICATION.equals(type) &&
                    json.has(Constants.JSON_KEY_NOTIFICATION_TITLE) &&
                    json.has(Constants.JSON_KEY_NOTIFICATION_MESSAGE) &&
                    json.has(Constants.JSON_KEY_NOTIFICATION_PRIORITY)) {

                String title = json.optString(Constants.JSON_KEY_NOTIFICATION_TITLE, "");
                String message = json.optString(Constants.JSON_KEY_NOTIFICATION_MESSAGE, "");
                String priority = json.optString(Constants.JSON_KEY_NOTIFICATION_PRIORITY, "Normal");

                // Ensure no null values are passed to constructor
                action = new com.example.autoflow.model.Action(type,
                        title.isEmpty() ? "Default Title" : title,
                        message.isEmpty() ? "Default Message" : message,
                        priority.isEmpty() ? "Normal" : priority);
            } else {
                // Create basic action with type only
                action = new com.example.autoflow.model.Action(type);
            }

            // Set value if present and not null
            if (json.has("value") && !json.isNull("value")) {
                String value = json.optString("value");
                if (value != null && !value.trim().isEmpty()) {
                    action.setValue(value.trim());
                }
            }

            return action;

        } catch (JSONException e) {
            System.err.println("Error parsing action JSON: " + e.getMessage());
            return null;
        } catch (IllegalArgumentException e) {
            System.err.println("Invalid action data: " + e.getMessage());
            return null;
        } catch (Exception e) {
            System.err.println("Unexpected error creating action: " + e.getMessage());
            return null;
        }
    }
    // Helper class for returning two strings (Java equivalent of Kotlin Pair)

    // Override equals and hashCode for proper object comparison
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        WorkflowEntity that = (WorkflowEntity) o;
        return id == that.id &&
                isEnabled == that.isEnabled &&
                Objects.equals(workflowName, that.workflowName) &&
                Objects.equals(triggerDetails, that.triggerDetails) &&
                Objects.equals(actionDetails, that.actionDetails);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, workflowName, isEnabled, triggerDetails, actionDetails);
    }

    @Override
    @NonNull
    public String toString() {
        return "WorkflowEntity{" +
                "id=" + id +
                ", workflowName='" + workflowName + '\'' +
                ", isEnabled=" + isEnabled +
                ", triggerDetails='" + triggerDetails + '\'' +
                ", actionDetails='" + actionDetails + '\'' +
                '}';
    }
}

File: app/src/main/java/com/example/autoflow/data/WorkflowEntity.kt
package com.example.autoflow.data

import android.util.Log
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import com.example.autoflow.model.Action
import com.example.autoflow.model.Trigger
import org.json.JSONArray
import org.json.JSONObject

@Entity(tableName = "workflows")
data class WorkflowEntity(
    @PrimaryKey(autoGenerate = true)
    var id: Long = 0,

    @ColumnInfo(name = "workflow_name")
    var workflowName: String = "",

    @ColumnInfo(name = "trigger_details")
    var triggerDetails: String = "",

    @ColumnInfo(name = "action_details")
    var actionDetails: String = "",

    @ColumnInfo(name = "is_enabled")
    var isEnabled: Boolean = true,

    @ColumnInfo(name = "trigger_logic")
    var triggerLogic: String = "AND",

    @ColumnInfo(name = "created_at")
    var createdAt: Long = System.currentTimeMillis(),

    @ColumnInfo(name = "updated_at")
    var updatedAt: Long = System.currentTimeMillis(),

    @ColumnInfo(name = "mode_id")
    var modeId: Long? = null,

    @ColumnInfo(name = "is_mode_workflow")
    var isModeWorkflow: Boolean = false
) {
    companion object {
        private const val TAG = "WorkflowEntity"

        fun fromTriggersAndActions(
            workflowName: String,
            isEnabled: Boolean,
            triggers: List<Trigger>,
            actions: List<Action>,
            triggerLogic: String = "AND"
        ): WorkflowEntity? {
            return try {
                val triggersJson = JSONArray()
                triggers.forEach { trigger ->
                    val triggerObj = JSONObject().apply {
                        put("type", trigger.type)
                        put("value", trigger.value)
                    }
                    triggersJson.put(triggerObj)
                }

                val actionsJson = JSONArray()
                actions.forEach { action ->
                    val actionObj = JSONObject().apply {
                        put("type", action.type)
                        action.value?.let { put("value", it) }
                        action.title?.let { put("title", it) }
                        action.message?.let { put("message", it) }
                        action.priority?.let { put("priority", it) }
                        action.duration?.let { put("duration", it) }
                    }
                    actionsJson.put(actionObj)
                }

                WorkflowEntity(
                    workflowName = workflowName,
                    triggerDetails = triggersJson.toString(),
                    actionDetails = actionsJson.toString(),
                    isEnabled = isEnabled,
                    triggerLogic = triggerLogic,
                    createdAt = System.currentTimeMillis(),
                    updatedAt = System.currentTimeMillis()
                )
            } catch (e: Exception) {
                Log.e(TAG, "❌ Error creating WorkflowEntity", e)
                null
            }
        }
    }
}

// ✅ EXTENSION FUNCTIONS - FIXED NULL SAFETY AND CONSTRUCTOR

fun WorkflowEntity.toTriggers(): List<Trigger> {
    return try {
        val triggersList = mutableListOf<Trigger>()
        val jsonArray = JSONArray(this.triggerDetails)

        for (i in 0 until jsonArray.length()) {
            val triggerObj = jsonArray.getJSONObject(i)
            val triggerType = triggerObj.getString("type")
            val triggerValue = triggerObj.optString("value", "")

            val trigger = when (triggerType.uppercase()) {
                "WIFI" -> {
                    try {
                        val valueJson = JSONObject(triggerValue)
                        val ssid = valueJson.optString("ssid", "")
                        val state = valueJson.optString("state", "CONNECTED")
                        Trigger.WiFiTrigger(ssid = ssid, state = state)
                    } catch (e: Exception) {
                        Trigger.WiFiTrigger(ssid = triggerValue, state = "CONNECTED")
                    }
                }
                "BLUETOOTH" -> {
                    try {
                        val valueJson = JSONObject(triggerValue)
                        val deviceAddress = valueJson.optString("deviceAddress", "")
                        val deviceName = valueJson.optString("deviceName", "")
                        Trigger.BluetoothTrigger(deviceAddress = deviceAddress, deviceName = deviceName)
                    } catch (e: Exception) {
                        Trigger.BluetoothTrigger(deviceAddress = triggerValue)
                    }
                }
                "LOCATION" -> {
                    try {
                        val valueJson = JSONObject(triggerValue)
                        val locationName = valueJson.optString("locationName", "Unknown Location")
                        val latitude = valueJson.optDouble("latitude", 0.0)
                        val longitude = valueJson.optDouble("longitude", 0.0)
                        val radius = valueJson.optDouble("radius", 100.0)
                        val triggerOnEntry = valueJson.optBoolean("triggerOnEntry", true)
                        val triggerOnExit = valueJson.optBoolean("triggerOnExit", false)
                        val triggerOn = valueJson.optString("triggerOn", "both")

                        Trigger.LocationTrigger(
                            locationName = locationName,
                            latitude = latitude,
                            longitude = longitude,
                            radius = radius,
                            triggerOnEntry = triggerOnEntry,
                            triggerOnExit = triggerOnExit,
                            triggerOn = triggerOn
                        )
                    } catch (e: Exception) {
                        Trigger.LocationTrigger(
                            locationName = "Unknown",
                            latitude = 0.0,
                            longitude = 0.0,
                            radius = 100.0,
                            triggerOnEntry = true,
                            triggerOnExit = false
                        )
                    }
                }
                "TIME" -> {
                    try {
                        val valueJson = JSONObject(triggerValue)
                        val time = valueJson.optString("time", "00:00")
                        val daysJsonArray = valueJson.optJSONArray("days")
                        val days = mutableListOf<String>()
                        if (daysJsonArray != null) {
                            for (j in 0 until daysJsonArray.length()) {
                                days.add(daysJsonArray.getString(j))
                            }
                        }
                        Trigger.TimeTrigger(time = time, days = days)
                    } catch (e: Exception) {
                        Trigger.TimeTrigger(time = triggerValue, days = emptyList())
                    }
                }
                "BATTERY" -> {
                    try {
                        val valueJson = JSONObject(triggerValue)
                        val level = valueJson.optInt("level", 20)
                        val condition = valueJson.optString("condition", "below")
                        Trigger.BatteryTrigger(level = level, condition = condition)
                    } catch (e: Exception) {
                        Trigger.BatteryTrigger(level = 20, condition = "below")
                    }
                }
                "MANUAL" -> {
                    Trigger.ManualTrigger(actionType = triggerValue.ifEmpty { "quick_action" })
                }
                else -> {
                    Trigger.ManualTrigger(actionType = "unknown")
                }
            }
            triggersList.add(trigger)
        }
        triggersList
    } catch (e: Exception) {
        Log.e("WorkflowEntity", "❌ Error parsing triggers", e)
        emptyList()
    }
}

fun WorkflowEntity.toActions(): List<Action> {
    return try {
        val actionsList = mutableListOf<Action>()
        val jsonArray = JSONArray(this.actionDetails)

        for (i in 0 until jsonArray.length()) {
            val actionObj = jsonArray.getJSONObject(i)
            val type = actionObj.getString("type")

            // Handle all optional fields safely
            val value = actionObj.optString("value", "")
            val title = actionObj.optString("title", "")
            val message = actionObj.optString("message", "")
            val priority = actionObj.optString("priority", "Normal")
            val duration = actionObj.optLong("duration", 0)

            // Added explicit null for scheduledUnblockTime to match signature if needed
            val action = Action(
                type = type,
                title = title,
                message = message,
                priority = priority,
                value = value,
                duration = duration,
                scheduledUnblockTime = null
            )
            actionsList.add(action)
        }
        actionsList
    } catch (e: Exception) {
        Log.e("WorkflowEntity", "❌ Error parsing actions", e)
        emptyList()
    }
}